##Data Preprocessing

In [ ]:
import numpy as np
import pandas as pd
import tensorflow as tf
tf.__version__

'2.8.0'

In [ ]:
df = pd.read_csv('/content/Churn_Modelling.csv')
df

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1          1               1   
1          1   83807.86              1          0               1   
2          8  159660.80              3          1               0   
3          1       0.00              2          0               0   
4          2  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

In [ ]:
x = df.iloc[:, 3:-1].values
y = df.iloc[:, -1].values
x

array([[619, 'France', 'Female', ..., 1, 1, 101348.88],
       [608, 'Spain', 'Female', ..., 0, 1, 112542.58],
       [502, 'France', 'Female', ..., 1, 0, 113931.57],
       ...,
       [709, 'France', 'Female', ..., 0, 1, 42085.58],
       [772, 'Germany', 'Male', ..., 1, 0, 92888.52],
       [792, 'France', 'Female', ..., 1, 0, 38190.78]], dtype=object)

In [ ]:
y

array([1, 0, 1, ..., 1, 1, 0])

In [ ]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
x[:, 2] = le.fit_transform(x[:, 2])
x

array([[619, 'France', 0, ..., 1, 1, 101348.88],
       [608, 'Spain', 0, ..., 0, 1, 112542.58],
       [502, 'France', 0, ..., 1, 0, 113931.57],
       ...,
       [709, 'France', 0, ..., 0, 1, 42085.58],
       [772, 'Germany', 1, ..., 1, 0, 92888.52],
       [792, 'France', 0, ..., 1, 0, 38190.78]], dtype=object)

In [ ]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoding', OneHotEncoder(), [1])], remainder='passthrough')
x = np.array(ct.fit_transform(x))
x

array([[1.0, 0.0, 0.0, ..., 1, 1, 101348.88],
       [0.0, 0.0, 1.0, ..., 0, 1, 112542.58],
       [1.0, 0.0, 0.0, ..., 1, 0, 113931.57],
       ...,
       [1.0, 0.0, 0.0, ..., 0, 1, 42085.58],
       [0.0, 1.0, 0.0, ..., 1, 0, 92888.52],
       [1.0, 0.0, 0.0, ..., 1, 0, 38190.78]], dtype=object)

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

In [ ]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test = sc.transform(x_test)

In [ ]:
x_train

array([[ 0.99476371, -0.57196155, -0.57869742, ...,  0.64706357,
        -1.03823026,  0.27554704],
       [ 0.99476371, -0.57196155, -0.57869742, ...,  0.64706357,
         0.96317747, -0.90640282],
       [ 0.99476371, -0.57196155, -0.57869742, ...,  0.64706357,
        -1.03823026,  1.57538739],
       ...,
       [ 0.99476371, -0.57196155, -0.57869742, ...,  0.64706357,
         0.96317747,  0.14679568],
       [-1.00526385, -0.57196155,  1.72801876, ..., -1.54544322,
        -1.03823026, -0.65775979],
       [-1.00526385,  1.74836928, -0.57869742, ...,  0.64706357,
         0.96317747, -0.2479495 ]])

#Building The ANN

###Initialize The ANN

In [ ]:
ann = tf.keras.models.Sequential()

###Adding Input Layer & First Hidden Layer

In [ ]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

we used 'relu' (rectified) activation function bcz in case of fully connected ann, we gotta use rectified activation function within the hidden layers

###Adding Second Hidden Layer

In [ ]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

###Adding Output Layer

In [ ]:
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

we used 'sigmoid' activation function bcz it gives us the output as well as the probablity of the output

#Traing The Ann

###Compiling The Ann

In [ ]:
ann.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

For binary classfication(yes or no 0/1), we use loss = 'binary_crossentropy' & activation='sigmoid' 

For multiclass classfication(east, west, north), we use loss = 'categorical_crossentropy' & activation='softmax' 


but in case of ANN regression we use loss = 'mean_squared_error' & no activation function

###Training ANN On The Training Set

In [ ]:
ann.fit(x_train, y_train, batch_size=32, epochs=100)

Epoch 1/100
250/250 [==============================] - 1s 1ms/step - loss: 0.5462 - accuracy: 0.7574
Epoch 2/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4638 - accuracy: 0.7986
Epoch 3/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4472 - accuracy: 0.8015
Epoch 4/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4398 - accuracy: 0.8051
Epoch 5/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4341 - accuracy: 0.8066
Epoch 6/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4293 - accuracy: 0.8100
Epoch 7/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4245 - accuracy: 0.8138
Epoch 8/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4195 - accuracy: 0.8156
Epoch 9/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4126 - accuracy: 0.8188
Epoch 10/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4029 - accura

here, batch_size defines the set of predicted values to be compared simultaniously with the actual output values

epochs means one complete forward & backward propagation of the data through the ANN

##Pedicting A Given Perticular Test Value

**Homework**

Use our ANN model to predict if the customer with the following informations will leave the bank: 

Geography: France

Credit Score: 600

Gender: Male

Age: 40 years old

Tenure: 3 years

Balance: \$ 60000

Number of Products: 2

Does this customer have a credit card ? Yes

Is this customer an Active Member: Yes

Estimated Salary: \$ 50000

So, should we say goodbye to that customer ?

In [ ]:
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])))

[[0.02358878]]


Therefore, our ANN model predicts that this customer stays in the bank!

**Important note 1:** Notice that the values of the features were all input in a double pair of square brackets. That's because the "predict" method always expects a 2D array as the format of its inputs. And putting our values into a double pair of square brackets makes the input exactly a 2D array.

**Important note 2:** Notice also that the "France" country was not input as a string in the last column but as "1, 0, 0" in the first three columns. That's because of course the predict method expects the one-hot-encoded values of the state, and as we see in the first row of the matrix of features X, "France" was encoded as "1, 0, 0". And be careful to include these values in the first three columns, because the dummy variables are always created in the first columns.

**Important note 3: As the training data was scalled, we need to scale the test input also

**Important note 4: As we used 'sigmoid' activation method in our output layer, so the predicted values is in the form of probablity

In [ ]:
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])) >0.5)

[[False]]


To convert the probablity into yes/no classification, we used a logic >0.5

##Predicting The Test Set Result

In [ ]:
np.set_printoptions(precision=2)

In [ ]:
y_pred = ann.predict(x_test)
y_pred = (y_pred>0.5)
print(np.concatenate((y_pred.reshape(len(y_pred), 1), y_test.reshape(len(y_test), 1)), axis=1))


[[0 0]
 [1 1]
 [0 0]
 ...
 [1 1]
 [0 0]
 [0 0]]


Here the terget 2 vectors i.e. y_pred & y_test are horizontal vectors. That's why we used axis=1 which means terget vectors are horizontal & we need to concatenate them into a vertical array

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
cm

array([[1516,   78],
       [ 198,  208]])

1516 = right prediction that stays in the bank

201 = right prediction that leaves the bank

198 = wrong prediction that stays in the bank

78 = wrong prediction leaves the bank

In [ ]:
accuracy_score(y_test, y_pred)

0.862